src/test/java/com/google/jenkins/plugins/storage/IT/ITUtil.java

```diff
@@ -0,0 +1,156 @@
+  }
+
+  /**
+   * Loads the content of the specified resource.
+   *
+   * @param testClass The test class the resource is being loaded for.
+   * @param name The name of the resource being loaded.
+   * @return The contents of the loaded resource.
+   * @throws IOException If an error occurred during loading.
+   */
+  static String loadResource(Class testClass, String name) throws IOException {
+    return new String(ByteStreams.toByteArray(testClass.getResourceAsStream(name)));
+  }
+
+  /**
+   * Dumps the logs from the specified {@link hudson.model.Run}.
+   *
+   * @param logger The {@link java.util.logging.Logger} to be written to.
+   * @param run The {@link hudson.model.Run} from which the logs will be read.
+   * @throws IOException If an error occurred while dumping the logs.
+   */
+  static void dumpLog(Logger logger, Run<?, ?> run) throws IOException {
+    BufferedReader reader = new BufferedReader(run.getLogReader());
+    String line = null;
+    while ((line = reader.readLine()) != null) {
+      logger.info(line);
+    }
+  }
+
+  /**
+   * Retrieves the location set through environment variables
+   *
+   * @return A Google Compute Resource Region (us-west1) or Zone (us-west1-a) string
+   */
+  static String getLocation() {
+    String location = System.getenv("GOOGLE_PROJECT_LOCATION");
+    if (location == null) {
+      location = System.getenv("GOOGLE_PROJECT_ZONE");
+    }
+    assertNotNull("GOOGLE_PROJECT_LOCATION env var must be set", location);
+    return location;
+  }
+
+  /**
+   * Given jenkins instance and credentials ID, return Credential
+   *
+   * @param itemGroup A handle to the Jenkins instance.
+   * @param credentialsId credentialsId to retrieve credential. Must exist in credentials store.
+   * @return Credential based on credentialsId.
+   * @throws GeneralSecurityException
+   */
+  static Credential getCredential(ItemGroup itemGroup, String credentialsId)
+      throws GeneralSecurityException {
+    GoogleRobotCredentials robotCreds =
+        CredentialsMatchers.firstOrNull(
+            CredentialsProvider.lookupCredentials(
+                GoogleRobotCredentials.class, itemGroup, ACL.SYSTEM, new ArrayList<>()),
+            CredentialsMatchers.withId(credentialsId));
+    try {
+      return robotCreds.getGoogleCredential(new StorageScopeRequirement());
+    } catch (GeneralSecurityException gse) {
+      throw new GeneralSecurityException(gse);
+    }
+  }
+
+  /**
+   * Given Jenkins instance and credentials ID, return a Storage object to make calls to the Google
+   * Cloud Storage JSON API.
+   *
+   * @param itemGroup A handle to the Jenkins instance.
+   * @param credentialsId credentialsId to retrieve credential. Must exist in credentials store.
+   * @return Storage object authenticated with credentialsId.
+   * @throws GeneralSecurityException
+   */
+  static Storage getService(ItemGroup itemGroup, String credentialsId)
```

Comment:
This should live in a client factory consistent with the other plugins approach.

Reply:
Right now, the entire plugin relies on a class (dependency of oauth plugin) called executor to execute storage API requests. It's acting like the client. Should I keep using the executor class in the IT's or start shifting over to a client factory approach.

Example of executor:
https://github.com/jenkinsci/google-storage-plugin/blob/631adc9c0724bd404012277cfe9e38087d8eb40f/src/main/java/com/google/jenkins/plugins/storage/AbstractUpload.java#L578-L583

Reply:
The executor class is providing retry logic on actions taken on an existing GS service client. My comment was specifically targeted at the creation of the GS service client itself.

More broadly speaking though, yes I think it would be worthwhile to ensure the code bases between plugins are consistent. I think it merits adding a TODO and linking an issue which calls for consolidation of client and retry logic into a client abstraction consistent with the other plugins.

Reply:
Done.